File: backend/orders/src/model/Order.kt
package challenge.model

import challenge.service.paymentstrategy.PaymentStrategy
import java.lang.Exception
import java.util.*

data class Order(val customer: Customer, val address: Address) {
    val items = mutableListOf<OrderItem>()
    val paymentStrategies = mutableListOf<PaymentStrategy>()
    var closedAt: Date? = null
        private set
    var payment: Payment? = null
    val totalAmount
        get() = items.sumByDouble { it.total }

    fun addProduct(product: Product, quantity: Int, paymentStrategy: PaymentStrategy) {
        val productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw Exception(DUPLICATED_PRODUCT_EXCEPTION_MESSAGE)

        items.add(OrderItem(product, quantity))
        paymentStrategies.add(paymentStrategy)
    }

    fun close() {
        closedAt = Date()
    }

    override fun toString(): String = "Items: " + this.items + " Total Price: " + totalAmount;

    companion object {
        const val DUPLICATED_PRODUCT_EXCEPTION_MESSAGE = "The product have already been added. Change the amount if you want more."
    }
}

File: backend/bootstrap/lib/bootstrap/order.rb
module BootStrap
  class Order
    attr_reader :customer,
                :items,
                :address,
                :closed_at

    def initialize(customer)
      @customer = customer
      @items = []
      @address = customer.address
    end

    def add_product(product)
      @items << product
    end

    def total_amount
      @items.map(&:price).inject(:+)
    end

    def close
      @closed_at = Time.now
    end
  end
end
